### ISSUE TITLE
Adapter fails to implement policy 

### ISSUE THREAD

**Issue by AdheipSingh** [apigee/istio-mixer-adapter#278]
Istio version used 1.1.4 , adapter version used 1.1.2. My Adapter is configured with Apigee and istio mixer and restricts the access to the  helloworld service without authentication in the default namespace. The bookinfo app works fine as in when tried to curl the /productpage i am restricted. 
However all other  microservices which i created are being able to be accessed without any authentication in the default namespace.(rule.yaml is enforcing policies in default namespace). Somehow my microservice are directly bypassing the adapter rules and are being directly being accessed.  Is there any specific attributes which i am missing or the anything to do with building the image ?
``` 
dell@kubernetes:~$ curl my-lb.com/productpage
PERMISSION_DENIED:apigee-handler.handler.istio-system:missing authentication
dell@kubernetes:~$ curl my-lb.com/hello
PERMISSION_DENIED:apigee-handler.handler.istio-system:missing authentication 
dell@kubernetes:~$ curl my-lb.com/api/cms
[{"id":"1","tool":"Ansible","languagebased":"Python","company":{"name":"RedHat","opensource":"AWX","certification":"RedHat Specialist Ansible Automation"}},{"id":"2","tool":"Puppet","languagebased":"Ruby","company":{"name":"None","opensource":"YES","certification":"RedHat Specialist Configuration Management"}}]
dell@kubernetes:~$ k get pods
NAME                              READY     STATUS    RESTARTS   AGE
api-6bd6b6567f-dnnzs              2/2       Running   0          14h
auth-6d799fc668-c6xwp             2/2       Running   0          20h
details-v1-8556db4b85-99s2h       2/2       Running   0          19h
helloworld-v1-85bbc4bcd9-mp5d6    2/2       Running   0          18h
helloworld-v2-6b4f6c8fd8-w8gbw    2/2       Running   0          18h
nginx-7c56fbdcbb-fj6ns            2/2       Running   0          18h
productpage-v1-66b8995967-nf6vv   2/2       Running   0          19h
ratings-v1-bb688c4fc-s49wr        2/2       Running   0          19h
reviews-v1-5f456964dd-p2df7       2/2       Running   0          19h
reviews-v2-84bd967859-prw8z       2/2       Running   0          19h
reviews-v3-5cb49b6b74-7mqln       2/2       Running   0          19h

```
This is the rules.yaml file
```---
apiVersion: config.istio.io/v1alpha2
kind: rule
metadata:
  name: apigee-rule
  namespace: istio-system
spec:
  match: context.reporter.kind == "inbound" && destination.namespace == "default" 
  actions:
  - handler: apigee-handler
    instances:
    - apigee-authorization
    - apigee-analytics

```

**Comment by theganyo** [comment]
Hmm. No, it shouldn't have anything do do with the image or attributes. If it works in some cases, it should work in similar cases unless there's something specifically different about that route. I'd start by checking a few things:

1. Did you add your api service to an Apigee Product?
2. Does the policy mixer log have any errors?
3. Does the adapter show a HandleAuthorization check when you attempt to access /api/cms?

**Comment by p6** [comment]
Answering on behalf of Adheip:

1. Yes. For Helloworld, it worked anyway for some reason (even if we deleted the helloworld product on Apigee) -- our assumption is that it should take over the whole namespace anyway. Is this not correct?
2. Not that I can see - here are the mixer debugger logs: https://gist.github.com/p6/9afd60ed6c10567eda54c0f980ade9ce
3. No

**Comment by p6** [comment]
To extend point #2 above FWIW - it seems that during the success path (using helloworld) - there are 4 entries in the mixer debug logs: two distinct context.reporter.kind = outbound (presumably from the ingress gateway and two entries of context.reporter.kind = inbound (this is presumably in the service pod sidecars).

In the failure path (using our services) - we only see two entries of context.reporter.kind = outbound. 

Related enquiry - is there a way to log + trace egress traffic from the adapter? I tried tracing the istio-auth proxy -- where I do see some inbound, but there's no real info in the req headers or body (empty) to correlate it to auth checks from the cluster.

**Comment by theganyo** [comment]
1. You're correct. Just asking for context.
2. I only see calls to "helloworld" (and "unknown") in the log, but nothing to your api service?

I'm not sure I understand your request about tracing adapter traffic.

**Comment by p6** [comment]
@theganyo  - thanks for the swift response

on 2. there is. you can see line 383 and 427 (my service is called "auth")

You can ignore the query on tracing adapter traffic. I was naively under the impression that the adapter calls out to management APIs on every call.. now i understand it syncs it at some frequency.

**Comment by theganyo** [comment]
Right. The adapter avoids affecting the request path as much as possible by maintaining local caches of control information.

Ok, so yes... I see your auth service. And yes, only outbound calls. Ok. This doesn't involve the adapter at all: It would appear that the auth sidecar just isn't doing any work. So, a couple more questions:

1. Is there an Istio sidecar installed properly on it? (`kubernetes describe pod ...`)
2. Is the sidecar envoy synced with Pilot? (`istioctl proxy-status`)

**Comment by AdheipSingh** [comment]
@theganyo  regarding 1. , here is how i am injecting the sidecar.
```
istioctl kube-inject -f name_of_deploymentfile.yaml | kubectl apply -f -
```
2. Here is an output of  istioctl  proxy status
```
istioctl proxy-status
NAME                                                  CDS        LDS        EDS               RDS        PILOT                            VERSION
api-6bd6b6567f-dnnzs.default                          SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
auth-6d59c6f454-twhvt.default                         SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
details-v1-8556db4b85-99s2h.default                   SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
helloworld-v1-75b7489bdc-x4z6j.default                SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
helloworld-v2-79666b8d86-m7hqc.default                SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
istio-ingressgateway-6bd68954d-zchd4.istio-system     SYNCED     SYNCED     SYNCED (100%)     SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
nginx-7c56fbdcbb-77qbq.default                        SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
productpage-v1-66b8995967-nf6vv.default               SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
ratings-v1-bb688c4fc-s49wr.default                    SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
reviews-v1-5f456964dd-p2df7.default                   SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
reviews-v2-84bd967859-prw8z.default                   SYNCED     SYNCED     SYNCED (50%)      SYNCED     istio-pilot-5f75bc4d45-mtqpt     1.1.3
reviews-v3-5cb49b6b74-7mqln.default                   SYNCED     SYNCED     SYNCED (50%)      SYNCED     

```

**Comment by theganyo** [comment]
Can you check another thing for me? Is the port for your service named "http" (or at least starts with "http-")?

`kubectl get service auth -o yaml`

If not, can you name the port "http" and try again?

**Comment by AdheipSingh** [comment]
After i changed the service file to http it worked. 
```
spec:
  clusterIP: 100.64.30.226
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 3000
  selector:
    app: auth
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

```

**Comment by theganyo** [comment]
Excellent! This is a recently discovered and fixed [bug](https://github.com/istio/istio/issues/13868) in Istio. It should be in the next release. I've also added a note on it in the wiki [here](https://github.com/apigee/istio-mixer-adapter/wiki/Debug-Authorization-check-in-Istio-1.1.x#is-istio-policy-support-enabled).

Sounds like you're good to go!

**Comment by AdheipSingh** [comment]
Thanks mate for the help  @theganyo

**Comment by duderino** [comment]
Fix is in Istio 1.1.7 which will be published in a few hours
